### ISSUE TITLE
ValueError: cannot handle a non-unique multi-index!

### ISSUE THREAD

**Issue by marberi** [dask/dask#4845]
[dask_error.txt](https://github.com/dask/dask/files/3220417/dask_error.txt)

This might be related to Dask not supporting multi-indexes. My code was randomly
failing, which made me first assume there was a problem in the input data. Running
with the versions

dask: 1.2.2
numpy: 1.16.3
pandas: 0.24.2

the minimal example below fails. Is there a way of making this error message more
intuitive? Or having this operation working.

"""
import numpy as np
import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame({'ind_a': np.arange(100), 'ind_b': 1, 'var': 'whatever'})
df = dd.from_pandas(df, npartitions=90)
    
# Only fails when grouping with two variables.
df['nr'] = df.groupby(['ind_a', 'ind_b']).cumcount()
len(df)
"""

**Comment by TomAugspurger** [comment]
Can you provide a reproducible example? http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by TomAugspurger** [comment]
Note that this only happens when we have multiple threads working with the same data.

```python

df['nr'] = df.groupby(['ind_a', 'ind_b']).cumcount()
len(df)


```python
In [4]: dask.config.set(scheduler='multiprocessing')
Out[4]: <dask.config.set at 0x113d8b208>

In [5]: df['nr'] = df.groupby(['ind_a', 'ind_b']).cumcount()
   ...: len(df)
   ...:
Out[5]: 100

In [6]: dask.config.set(scheduler='single-threaded')
Out[6]: <dask.config.set at 0x113d9ae80>

In [7]: df['nr'] = df.groupby(['ind_a', 'ind_b']).cumcount()
   ...: len(df)
```

Seems like some piece of data is being manipulated inplace by two threads.

**Comment by jakirkham** [comment]
If that's true, would it be possible to inject copies in-between steps to identify where the in-place manipulation is occurring?

**Comment by TomAugspurger** [comment]
I'm not sure. I think a more productive line is to demonstrate the thread safety issue with pandas-only code and a ThreadPool, and see if changes can be made to pandas. @marberi are you interested in exploring that?

**Comment by marberi** [comment]
@TomAugspurger Could you outline how I could make such a test? I am quite familiar with Pandas from as a user, but not how one would debug this issue.

**Comment by TomAugspurger** [comment]
I'm not entirely sure either. This is a bit tricky, since it's not clear
what section of pandas is mutating the shared object.

I would try a couple approaches:

1. Use a ThreadPoolExecutor to apply multiple reductions (cumcount, commix)
to the same DataFrameGroupBy object at the same time.
   That would help narrow down whether the issue is in pandas' groupby
code, or in the indexing dask does before getting there.
2. Try adding finer and finer-grained thread locks on the functions used as
part of dask's cumcount.

This still may not end up finding this issue... This is a tricky one.

**Comment by jakirkham** [comment]
@marberi, does Tom's suggested course of action above make sense? Do you have additional questions?

**Comment by marberi** [comment]
I will have a try. Tried some simple debugging this morning, inserting some print statements in the Pandas code to see if I could find some hints. Quite tricky, since the state at times changes between "is_unique" and checking what are duplicates. I keep looking. However, testing if upgrading Pandas to the latest version gives another issue (see attachment). Are you aware of this problem? It seems unrelated, but something which should be fixed.
[new_message.txt](https://github.com/dask/dask/files/3252464/new_message.txt)

**Comment by marberi** [comment]
Thanks. It looks very relevant. The Pandas code called by dask crashes in multiple places, but testing with "is_unique" seems to always be involved.

**Comment by TomAugspurger** [comment]
Yes, I missed that earlier. You might try adding a `threading.Lock` around
uses of is_unique / do_unique_check just to verify.

**Comment by marberi** [comment]
I tried adding a lock in "is_unique" in the Pandas source code, as shown below. Then
rerunning I still have the same issue. Does this look like a correct usage of locking?

```
    @property
    def is_unique(self):
        import threading
        lock = threading.Lock()
        lock.acquire_lock()

        print('is rebuilt: 2')
        if self.need_unique_check:
            self._do_unique_check()

        ans = self.unique == 1
        lock.release()

        return ans
```

**Comment by TomAugspurger** [comment]
Yes. Pandas has multiple index classes, and some override is_unique with
their own implementations.

**Comment by marberi** [comment]
Ok, I got one step further concerning this. The "is_unique" method in the Index class found in:
pandas/pandas/core/indexes/base.py

has a "@cache_readonly" decorator. When commenting out this cache, this problem disappears. This appears to be a custom decorator implemented in the Pandas code.

**Comment by TomAugspurger** [comment]
Yeah, the output of `is_unique` is computed once and cached for the
lifetime of the object. Future calls to `is_unique` won't even reach the
lock, after the value has been cached.

**Comment by marberi** [comment]
Would this be an acceptable fix? I tested acquiring locks inside the __get__ method itself and
it did not work. 

https://github.com/marberi/pandas/commit/ad33858298961d63ab0bc70caf1d04b3ca02b5fc

**Comment by TomAugspurger** [comment]
May be worth posting on the pandas issue I linked to. I'm not familiar with
pandas' indexing code.

**Comment by marberi** [comment]
Ok, also posted there. Let's see what they say.

**Comment by jakirkham** [comment]
Is this still an issue, @marberi?

**Comment by marberi** [comment]
@jakirkham 

Creating a new environment with:
dask: 2.20.0
numpy: 1.18.5
pandas: 1.0.5

I can confirm the error is still there. Testing this last year, we could localize
where the problem came from, but without determining the best way to 
develop a fix.

**Comment by marberi** [comment]
Reporting another bug, I wanted to test if the problem reported here still is an issue. The code in the original 
report (https://github.com/dask/dask/issues/4845#issue-448576561) appears to work fine using Dask version 
2022.9.1 (Linux, Python 3.10.6, Conda). Could someone else confirm this?

**Comment by jakirkham** [comment]
Yeah am no longer seeing the exception in the original example. Maybe it is worth adding this as a test to Dask (to protect against regression). Would this be of interest to you @marberi?

**Comment by marberi** [comment]
@jakirkham Added a test here:
https://github.com/dask/dask/pull/9506

as mentioned, when running "pytest" I could not get it to run the new test. It might
be something trivial. Tests are still running.

**Comment by marberi** [comment]
Not from my side. Hopefully this gets fixed.

**Comment by phofl** [comment]
@Giatroo do you have a reproducer? The initial example works now

**Comment by Giatroo** [comment]
The initial example still doesn't work. You just have to make a few tries, since it's an error that appears randomly. 

Here's a print of an env I've just created by installing `dask[dataframe]` on the latest version.

![image](https://github.com/dask/dask/assets/13424358/1cd58988-58cd-4fb0-9a44-030b809860c6)
